src/Enzo/enzo_EnzoMethodPmDeposit.cpp

```diff
@@ -97,221 +105,271 @@ void EnzoMethodPmDeposit::compute ( Block * block) throw()
-	     (ba == be));
+    int num_mass = particle_groups->size("has_mass");
 
     // Accumulate particle density using CIC
 
     enzo_float cosmo_a=1.0;
     enzo_float cosmo_dadt=0.0;
     EnzoPhysicsCosmology * cosmology = enzo::cosmology();
-    
+
     if (cosmology) {
 
       double time = block->time();
       double dt   = block->dt();
       cosmology->compute_expansion_factor (&cosmo_a,&cosmo_dadt,time+alpha_*dt);
-      
+
     }
     if (rank >= 1) hx *= cosmo_a;
     if (rank >= 2) hy *= cosmo_a;
     if (rank >= 3) hz *= cosmo_a;
-    
+
     const double dt = alpha_ * block->dt() / cosmo_a;
 
-    // Scale mass by volume if particle value is mass instead of density
+    const int level = block->level();
 
-    int level = block->level();
+    // Loop over all particles that have mass
+    for (int ipt = 0; ipt < num_mass; ipt++){
 
-    dens *= std::pow(2.0,rank*level);
+      const int it = particle.type_index( particle_groups->item("has_mass",ipt));
 
-    // Accumulated single velocity array for Baryon deposit
+      // check precisions match
 
-    for (int ib=0; ib<particle.num_batches(it); ib++) {
+      const int ia = particle.attribute_index(it,"x");
+      const int ba = particle.attribute_bytes(it,ia); // "bytes (actual)"
+      const int be = sizeof(enzo_float);                // "bytes (expected)"
 
-      const int np = particle.num_particles(it,ib);
+      ASSERT4 ("EnzoMethodPmUpdate::compute()",
+               "Particle type %s attribute %s defined as %s but expecting %s",
+               particle.type_name(it).c_str(),
+               particle.attribute_name(it,ia).c_str(),
+               ((ba == 4) ? "single" : ((ba == 8) ? "double" : "quadruple")),
+               ((be == 4) ? "single" : ((be == 8) ? "double" : "quadruple")),
+               (ba == be));
 
-      if (rank == 1) {
 
-	const int ia_x  = particle.attribute_index(it,"x");
-	const int ia_vx = particle.attribute_index(it,"vx");
+      enzo_float dens = 0.0;
+      int ia_m = -1;
 
-	enzo_float * xa =  (enzo_float *)particle.attribute_array (it,ia_x,ib);
-	enzo_float * vxa = (enzo_float *)particle.attribute_array (it,ia_vx,ib);
+      // Determine if particle has a constant mass, or if mass varies
+      // and we need to use the mass attribute
+      if (particle.has_constant (it,"mass")){
 
-	const int dp =  particle.stride(it,ia_x);
-	const int dv =  particle.stride(it,ia_vx);
+        const int ic_mass = particle.constant_index (it, "mass");
+        dens = *((enzo_float *)(particle.constant_value (it,ic_mass)));
+      } else {
+        dens = 1.0;
+        ia_m = particle.attribute_index(it, "mass");
+      }
 
-	for (int ip=0; ip<np; ip++) {
+      // Scale mass by volume if particle value is mass instead of density
+      // Required for Cosmology ("mass" is mass)
+      // Not for Collapse ("mass" is density)
+      if (cosmology) {
+        dens *= std::pow(2.0,rank*level);
+      }
```

Comment:
This is the bit I'm not sure about. I had thought particle masses were stored as densities in all cases (at least this is the case for Enzo).... So is this really a cosmology vs. non-cosmology thing or just a quirk of the collapse test?

Reply:
@jobordner can you comment on this, please?